presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -570,6 +571,11 @@ public PlanOptimizers(
                         .add(new InlineProjections(metadata.getFunctionAndTypeManager()))
                         .build()));
 
+        // Cache history based plan hashes for each plan node.
+        // After this step, nodes with same PlanNodeId will share same history based statistics.
+        // TODO: Enforce that no rules after this change plan node ids unless needed
```

Comment:
I have several questions about this:
1. What is the purpose of caching the statistics at this point?
2.  what does "unless needed" mean.  Like what happens to HBO if we decide something "needs" to change plan node ids?
3. This is a significant restriction on how optimizations after this point can work.  Is the benefit worth the cost?
4. This seems very fragile in practice.   Any optimizer might change plan node ids and it could easily be missed in review.  If this is important, it needs an enforcement mechanism (plan validator, tests, etc.)  to actually prevent this from happening.

Reply:
Discussed offline